firewood/src/db.rs

```diff
@@ -301,17 +306,76 @@ impl<S: ShaleStore<Node> + Send + Sync> api::DbView for DbRev<S> {
             .map_err(|e| api::Error::IO(std::io::Error::new(ErrorKind::Other, e)))
     }
 
-    async fn range_proof<K: api::KeyType, V, N>(
+    async fn range_proof<K: api::KeyType, V>(
         &self,
-        _first_key: Option<K>,
-        _last_key: Option<K>,
-        _limit: usize,
-    ) -> Result<Option<api::RangeProof<K, V, N>>, api::Error> {
-        todo!()
+        first_key: Option<K>,
+        last_key: Option<K>,
+        limit: usize,
+    ) -> Result<Option<api::RangeProof<Vec<u8>, Vec<u8>>>, api::Error> {
+        let mut it = self.stream(first_key.as_ref())?;
+        let first_key = match it.next().await {
+            None => return Ok(None),
+            Some(Err(e)) => return Err(e),
+            Some(Ok((key, _))) => self.single_key_proof(key).await?,
+        };
+        let Some(first_key) = first_key else {
+            return Ok(None);
+        };
+        let mut found = 0;
+
+        // middle is mutable because the last key will be removed before returning
+        let mut middle = it
+            .take_while(|kv| {
+                future::ready({
+                    // we stop if this returns true. We return true if:
+                    // (a) (a1) we asked for a key count limit AND
+                    //     (a2) we reached the limit
+                    // OR
+                    // (b) (b1) we specified a stop key AND
+                    //     (b2) the key returned is after the specified key; AND
+                    //     (b3) the Result returned wasn't an error
+                    found += 1;
+                    (limit > 0 && found > limit) // (a)
+                        || (last_key.is_some() // (b1)
+                            && kv
+                                .as_ref()
+                                .map(|(k, _)| {
+                                    <Vec<u8> as AsRef<[u8]>>::as_ref(k)
+                                        < last_key.as_ref().unwrap().as_ref()
+                                }) // (b2)
+                                .unwrap_or(false)) // (b3)
+                })
+            })
+            .try_collect::<Vec<(Vec<u8>, Vec<u8>)>>()
```

Comment:
I think what you really want is:
```rust
        let mut middle = stream
            .take_while(|kv_result| {
                let Some(last_key) = last_key.as_ref() else {
                    return ready(true);
                };

                let result = kv_result
                    .as_ref()
                    .map(|(key, _)| key.as_slice() < last_key.as_ref())
                    .unwrap_or(true); // always return Err(_)

                ready(result)
            })
            .take(limit)
            .try_collect::<Vec<_>>()
            .await?;
```

Reply:
I refactored this. Take should probably be first, since it's way cheaper.

Reply:
Much better